core/src/main/scala/org/apache/spark/sql/delta/stats/PrepareDeltaScan.scala

```diff
@@ -0,0 +1,338 @@
+      OptimisticTransaction.getActive.foreach { txn =>
+        val logsInPlan = plan.collect { case DeltaTable(fileIndex) => fileIndex.deltaLog }
+        if (logsInPlan.exists(_.isSameLogAs(txn.deltaLog))) {
+          txn.readWholeTable()
+        }
+      }
+
+      // Just return the plan if statistics based skipping is off.
+      // It will fall back to just partition pruning at planning time.
+      plan
+    }
+  }
+
+  object DeltaTableScan {
+
+    /**
+     * The components of DeltaTableScanType are:
+     * - an `AttributeSet` of the project collected by `PhysicalOperation`
+     * - filter expressions collected by `PhysicalOperation`
+     * - the `TahoeLogFileIndex` of the matched DeltaTable`
+     * - integer value of limit expression, if any
+     * - matched `DeltaTable`
+     */
+    private type DeltaTableScanType =
+      (Seq[Attribute], Seq[Expression], TahoeLogFileIndex, Option[Int], LogicalRelation)
+
+    def unapply(plan: LogicalPlan): Option[DeltaTableScanType] = {
+      val limitPushdownEnabled = spark.conf.get(DeltaSQLConf.DELTA_LIMIT_PUSHDOWN_ENABLED)
+      plan match {
+        case LocalLimit(IntegerLiteral(limit),
+          PhysicalOperation(project, filters, delta @ DeltaTable(fileIndex: TahoeLogFileIndex)))
+            if limitPushdownEnabled && containsPartitionFiltersOnly(filters, fileIndex) =>
+          val projects = AttributeSet(project).toSeq
+          Some((projects, filters, fileIndex, Some(limit), delta))
+
+        case PhysicalOperation(
+            project,
+            filters,
+            delta @ DeltaTable(fileIndex: TahoeLogFileIndex)) =>
+          val projects = AttributeSet(project).toSeq
+          val allFilters = fileIndex.partitionFilters ++ filters
+          Some((projects, allFilters, fileIndex, None, delta))
+
+        case _ => None
+      }
+    }
+
+    private def containsPartitionFiltersOnly(
+        filters: Seq[Expression],
+        fileIndex: TahoeLogFileIndex): Boolean = {
+      val partitionColumns = fileIndex.deltaLog.snapshot.metadata.partitionColumns
+      import DeltaTableUtils._
+      filters.forall(expr => !containsSubquery(expr) &&
+        isPredicatePartitionColumnsOnly(expr, partitionColumns, spark))
+    }
+  }
+}
+
+class PrepareDeltaScan(protected val spark: SparkSession)
+  extends PrepareDeltaScanBase
+
+object PrepareDeltaScanBase {
+
+  /**
+   * Optional callback function that is called after `getDeltaScanGenerator` is called
+   * by the PrepareDeltaScan rule. This is primarily used for testing purposes.
+   */
+  @volatile private var onGetDeltaScanGeneratorCallback: DeltaScanGenerator => Unit = _
+
+  /**
+   * Run a thunk of code with the given callback function injected into the PrepareDeltaScan rule.
+   * The callback function is called after `getDeltaScanGenerator` is called
+   * by the PrepareDeltaScan rule. This is primarily used for testing purposes.
+   */
+  private[delta] def withCallbackOnGetDeltaScanGenerator[T](
```

Comment:
i dont see this being used anywhere. is this something your next PR will have that made its way here? if so please remove it. :)